exercises/simple-cipher/tests/simple-cipher.rs

```diff
@@ -0,0 +1,136 @@
+extern crate simple_cipher;
+use simple_cipher::*;
+use std::collections::HashSet;
+
+const PLAIN_TEXT: &str = "thisismysecret";
+const KEY: &str = "abcdefghij";
+
+
+#[test]
+fn cipher_can_encode_with_given_key() {
+    assert_eq!(encode(KEY, "aaaaaaaaaa"), Some(KEY.to_string()));
+}
+
+#[test]
+#[ignore]
+fn cipher_can_decode_with_given_key() {
+    assert_eq!(decode(KEY, "abcdefghij"), Some("aaaaaaaaaa".to_string()));
+}
+
+#[test]
+#[ignore]
+fn cipher_is_reversible_given_key() {
+    assert_eq!(
+        decode(KEY, &encode(KEY, PLAIN_TEXT).unwrap()),
+        Some(PLAIN_TEXT.to_string())
+    );
+}
+
+#[test]
+#[ignore]
+fn cipher_can_double_shift_encode() {
+    let plain_text = "iamapandabear";
+    assert_eq!(
+        encode(plain_text, plain_text),
+        Some("qayaeaagaciai".to_string())
+    );
+}
+
+#[test]
+#[ignore]
+fn cipher_can_wrap_encode() {
+    assert_eq!(encode(KEY, "zzzzzzzzzz"), Some("zabcdefghi".to_string()));
+}
+
+#[test]
+#[ignore]
+fn cipher_can_encode_a_message_that_is_shorter_than_the_key() {
+    assert_eq!(encode(KEY, "aaaaa"), Some("abcde".to_string()));
+}
+
+#[test]
+#[ignore]
+fn cipher_can_decode_a_message_that_is_shorter_than_the_key() {
+    assert_eq!(decode(KEY, "abcde"), Some("aaaaa".to_string()));
+}
+
+#[test]
+#[ignore]
+fn encode_returns_none_with_an_all_caps_key() {
+    let key = "ABCDEF";
+    assert_eq!(encode(key, PLAIN_TEXT), None);
+}
+
+#[test]
+#[ignore]
+fn encode_returns_none_with_an_any_caps_key() {
+    let key = "abcdEFg";
+    assert_eq!(encode(key, PLAIN_TEXT), None);
+}
+
+#[test]
+#[ignore]
+fn encode_returns_none_with_numeric_key() {
+    let key = "12345";
+    assert_eq!(encode(key, PLAIN_TEXT), None);
+}
+
+#[test]
+#[ignore]
+fn encode_returns_none_with_any_numeric_key() {
+    let key = "abcd345ef";
+    assert_eq!(encode(key, PLAIN_TEXT), None);
+}
+
+#[test]
+#[ignore]
+fn encode_returns_none_with_empty_key() {
```

Comment:
clarify: should it also return `None` with a key that is non-empty but shorter than the message? If so, is there a test for that? (I may have missed it in my haste, so please excuse my inattentiveness)

What about the same tests for decode?

Reply:
A key that is shorter than message is fine. A one character key would be equivalent to a Caesar cipher. The default key in the tests is shorter than the message so we do test that case. 

All these return none tests are only testing the encode function, but yes, it would be better to test both encode and decode. Before I add this, is there a preference for adding a test for each encode/decode case or combining the assert statements in one test?

Reply:
> Before I add this, is there a preference for adding a test for each encode/decode case or combining the assert statements in one test?

I am thinking it would be better to have them be separate, and my reason is this: If both asserts were to fail, this makes it possible to determine which one failed from just the test name, instead of having to look at the line number.

I wrote than and then realised that even if they two asserts were combined in one test, it may be possible to determine which one failed from reading the test failure message. In that case that makes my preference weaker, though I guess it still exists.